Cabal/src/Distribution/PackageDescription/Check/Common.hs

```diff
@@ -116,34 +122,58 @@ partitionDeps ads ns ds = do
 -- for important dependencies like base).
 checkPVP
   :: Monad m
-  => (String -> PackageCheck) -- Warn message dependend on name
+  => (Dependency -> Bool)
+  -> (String -> PackageCheck) -- Warn message dependend on name
   -- (e.g. "base", "Cabal").
   -> [Dependency]
   -> CheckM m ()
-checkPVP ckf ds = do
-  let ods = checkPVPPrim ds
+checkPVP p ckf ds = do
+  let ods = filter p ds
   mapM_ (tellP . ckf . unPackageName . depPkgName) ods
 
 -- PVP dependency check for a list of dependencies. Some code duplication
 -- is sadly needed to provide more ergonimic error messages.
 checkPVPs
   :: Monad m
-  => ( [String]
+  => (Dependency -> Bool)
+  -> ( [String]
        -> PackageCheck -- Grouped error message, depends on a
        -- set of names.
      )
   -> [Dependency] -- Deps to analyse.
   -> CheckM m ()
-checkPVPs cf ds
+checkPVPs p cf ds
   | null ns = return ()
   | otherwise = tellP (cf ns)
   where
-    ods = checkPVPPrim ds
+    ods = filter p ds
     ns = map (unPackageName . depPkgName) ods
 
--- Returns dependencies without upper bounds.
-checkPVPPrim :: [Dependency] -> [Dependency]
-checkPVPPrim ds = filter withoutUpper ds
+-- | Is the version range without an upper bound?
+withoutUpperBound :: Dependency -> Bool
+withoutUpperBound (Dependency _ ver _) = not . hasUpperBound $ ver
+
+-- | Is the upper bound version range LEQ (less or equal, <=)?
+leqUpperBound :: Dependency -> Bool
+leqUpperBound (Dependency _ ver _)
```

Comment:
Have you considered moving this to `Distribution.Types.Dependency`?

Not a request, just want to know what your thoughts about it are.